components/rollup_node/block_gen.md

```diff
@@ -0,0 +1,147 @@
+
+The types of deposit blocks are as follows:
+
+```python
+class DepositBlock:
+    deposits: List[Deposit]
+
+class Deposit:
+    feedIndex: uint64
+    GasLimit:  uint64
+
+class UserDeposit(Deposit):
+    isEOA:       bool
+    l1TxOrigin:  Address
+    target:      Address
+    data:        bytes
+
+class ContextDeposit(Deposit):
+    blockHash:   bytes32
+    blockNumber: uint64
+    timestamp:   uint64
+    baseFee:     uint64
+```
+
+### Deposit Block Generation
+
+The function `generate_deposit_block(l1_block_number)` is defined as:
+
+- Generate the `ContextDeposit` using the L1 block body. Specifically the L1 block's `blockHash`, `blockNumber`, `timestamp`, and `baseFee`.
+- Get all events emitted by the `DepositFeed` contract at block `l1_block_number`.
+    - For each event, generate a new `UserDeposit` based on the emitted calldata. All fields should be emitted by DepositFeed contract.
+
+#### **Footnote**: The DepositFeed ensures that no matter the size of the event it is possible to prove the deposit data to the fraud proof. See the DepositFeed spec for details.
+
+## Sequencer Blocks in Depth
+
+Sequencer blocks contain a majority of the transactions submitted to the rollup chain. Sequencer blocks can also influence the current timestamp of L2 which is exposed to contracts & is used in the EIP1559 fee pricing calculations.
+
+Sequencer blocks are of the following form:
+
+```python
+class SequencerBlock:
+    target_epoch: uint64  # epoch number that this block is intended for
+    sequencer_suggested_timestamp: uint64  # a timestamp suggested by the sequencer
+    transactions: List[Transaction]  # A list of transactions in the block
+```
+
+These sequencer blocks must be validated as properly formatted. These validity checks are:
+
+1. `target_epoch` must be within the sequencing window (introduced in the previous sections)
+2. `sequencer_suggested_timestamp` must be monotonic & within an acceptable range.
+3. `transactions` must all be correctly formatted (eg. contain valid signatures).
+
+If any of these checks fail the block must be skipped or ignored. Otherwise, all transactions in the sequencer block will be executed.
+
+## Timestamps in Depth
+
+Time can be manipulated by the sequencer; however, it must stay within reasonable bounds and must be monotonic (ie. can never go backwards). The specific conditions that must be met are:
+
+```python
+# last_l1_timestamp is the L1 timestamp set during the last ContextDeposit
+# last_timestamp is the timestamp from the previous rollup block
+# average_l1_block_time is the average time between two L1 blocks
+# sequencing_window_size is the number of L1 blocks the sequencer has to submit their blocks
+min_timestamp = max(last_l1_timestamp, last_timestamp)
+max_timestamp = last_l1_timestamp + sequencing_window_size * average_l1_block_time
```

Comment:
We should be sure that this is safe when it comes to post-merge. If i understand correctly, the blocktime is basically deterministic and will effectively double if a slot is missed. I think it will be, but potentially good to note here eventually

Reply:
I don't think it will be an issue because the actual 'max timestamp' is independent of the L1 block time. I just used the L1 block time as the config because otherwise I would have to introduce another "window", this time a "suggested timestamp window". So the updated equation would be:

```
max_timestamp = last_l1_timestamp + suggested_timestamp_window_size
```

So the way that I wrote it is a little misleading but writing it this way introduces another concept which scared me a bit.

I'm super interested in feedback on:
1. Which approach you like better
2. If there's a better variable name for this than `suggested_timestamp_window_size` 😅😭